src/rpcserver.cpp

```diff
@@ -118,25 +119,21 @@ void RPCTypeCheckObj(const UniValue& o,
     }
 }
 
-static inline int64_t roundint64(double d)
-{
-    return (int64_t)(d > 0 ? d + 0.5 : d - 0.5);
-}
-
 CAmount AmountFromValue(const UniValue& value)
 {
-    double dAmount = value.get_real();
-    if (dAmount <= 0.0 || dAmount > 21000000.0)
-        throw JSONRPCError(RPC_TYPE_ERROR, "Invalid amount");
-    CAmount nAmount = roundint64(dAmount * COIN);
-    if (!MoneyRange(nAmount))
+    if (!value.isReal() && !value.isNum())
+        throw JSONRPCError(RPC_TYPE_ERROR, "Amount is not a number");
+    CAmount amount;
+    if (!ParseMoney(value.getValStr(), amount))
         throw JSONRPCError(RPC_TYPE_ERROR, "Invalid amount");
-    return nAmount;
+    if (!MoneyRange(amount))
```

Comment:
I just noticed that there is a slight change of semantics here. The old check `dAmount <= 0.0 || dAmount > 21000000.0` did not allow zero amounts, whereas `MoneyRange` allows `nValue >= 0 && nValue <= MAX_MONEY`.

Unsure what to do here. The no-brainer solution would be to restore the old behavior, however 
- There is one place in the source where this is worked around with an ugly hack: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcwallet.cpp#L2172
- The other places where AmountFromValue is used are `sendtoaddress` `movecmd` `sendfrom` `sendmany`

The alternative would be to handle zero-ness at the other call sites, which is feasible as they are few in number.